feat(coderd/agentapi): allow inserting apps for sub agents

DanielleMaywood opened the pull request:
Relates to https://github.com/coder/internal/issues/614

Introduces a new field to the `CreateSubAgent` request. This new field allows adding workspace apps to a sub agent at the point of creation.

DanielleMaywood committed:
feat(coderd/agentapi): allow inserting workspace apps for subagents

Allow the creation of workspace apps when creating a sub agent.

DanielleMaywood committed:
chore: revert dbauthz change

I think the dbauthz change is out-of-scope for this change

DanielleMaywood commented on agent/proto/agent.proto:
```diff
@@ -388,6 +388,41 @@ message CreateSubAgentRequest {
 	string architecture = 3;
 	string operating_system = 4;
+
+	message App {
+		message Healthcheck {
+			int32 interval = 1;
+			int32 threshold = 2;
+			string url = 3;
+		}
+
+		enum OpenIn {
+			SLIM_WINDOW = 0;
+			TAB = 1;
+		}
+
+		enum Share {
+			OWNER = 0;
+			AUTHENTICATED = 1;
+			PUBLIC = 2;
+		}
+
+		string slug = 1;
+		optional string command = 2;
+		optional string display_name = 3;
+		optional bool external = 4;
+		optional string group = 5;
+		optional Healthcheck healthcheck = 6;
+		optional bool hidden = 7;
+		optional string icon = 8;
+		optional OpenIn open_in = 9;
+		optional int32 order = 10;
+		optional Share share = 11;
+		optional bool subdomain = 12;
+		optional string url = 13;
+	}
```
I'm aware we already have a `WorkspaceApp` structure defined in `agent.proto`, but that is used as part of the `Manifest` response structure.

The reasoning for creating a new structure is that the existing `WorkspaceApp` is for an existing workspace app, whereas this new version is for requesting to make a _new_ workspace app.

The design of this structure is based on the terraform provider `coder_app` resource.

DanielleMaywood commented on coderd/agentapi/subagent.go:
```diff
@@ -71,6 +72,73 @@ func (a *SubAgentAPI) CreateSubAgent(ctx context.Context, req *agentproto.Create
+			openIn = database.WorkspaceAppOpenInTab
+		}
+
+		//nolint:gocritic // We've already inserted the agent with restricted permissions, so we're safe to do this.
+		_, err := a.Database.InsertWorkspaceApp(dbauthz.AsSystemRestricted(ctx), database.InsertWorkspaceAppParams{
```
I'm not a fan of `AsSystemRestricted`. I was tempted to update the `dbauthz` implementation of `InsertWorkspaceApp` but I feel that might be out-of-scope for this PR.

DanielleMaywood replied:
Have added the TODO and have created an issue https://github.com/coder/coder/issues/18210

DanielleMaywood replied:
A PR is up https://github.com/coder/coder/pull/18223

spikecurtis replied:
I added the sub-task. Was hoping for an "ack", but seems this is now moot. 👍

DanielleMaywood committed:
chore: appease linter

DanielleMaywood committed:
chore: add some more tests

DanielleMaywood committed:
Merge branch 'main' into dm-sub-agent-workspace-apps

DanielleMaywood committed:
Merge branch 'main' into dm-sub-agent-workspace-apps

DanielleMaywood committed:
test: workspace apps are deleted

DanielleMaywood committed:
Merge branch 'main' into dm-sub-agent-workspace-apps

johnstcn commented on coderd/agentapi/subagent.go:
```diff
@@ -71,6 +72,72 @@ func (a *SubAgentAPI) CreateSubAgent(ctx context.Context, req *agentproto.Create
 	}
 
+	for _, app := range req.Apps {
+		slug := app.Slug
+		if slug == "" {
+			return nil, xerrors.Errorf("app must have a slug or name set")
+		}
+		if !provisioner.AppSlugRegex.MatchString(slug) {
+			return nil, xerrors.Errorf("app slug %q does not match regex %q", slug, provisioner.AppSlugRegex)
+		}
+
+		health := database.WorkspaceAppHealthDisabled
+		if app.Healthcheck == nil {
+			app.Healthcheck = &agentproto.CreateSubAgentRequest_App_Healthcheck{}
+		}
+		if app.Healthcheck.Url != "" {
+			health = database.WorkspaceAppHealthInitializing
+		}
+
+		sharingLevel := database.AppSharingLevelOwner
+		switch app.GetShare() {
+		case agentproto.CreateSubAgentRequest_App_AUTHENTICATED:
+			sharingLevel = database.AppSharingLevelAuthenticated
+		case agentproto.CreateSubAgentRequest_App_PUBLIC:
+			sharingLevel = database.AppSharingLevelPublic
+		}
+
+		openIn := database.WorkspaceAppOpenInSlimWindow
+		if app.GetOpenIn() == agentproto.CreateSubAgentRequest_App_TAB {
+			openIn = database.WorkspaceAppOpenInTab
+		}
```
If the validation logic is the same for both `provisionerd` and for this use-case, could we extract it somewhere so it is kept the same in both places?

DanielleMaywood replied:
Possibly, although we use different types so the only shared logic would be checking for an empty name or to match the regex. I'm happy to either keep the current approach, or extract it out, but with a preference for the current approach.

johnstcn reviewed (commented):
Because you add new types here, we should also update the comments in `tailnet/proto/version.go` to reflect this. It hasn't been shipped yet AFAIK so it should be OK to just keep the same version.

The other potential issue I see here is correctly surfacing validation errors from the `CreateSubAgent()` calls to the user. I don't see any calls to this yet, but we'll want to make sure that users are able to correctly address issues that cause validation to fail.

DanielleMaywood committed:
refactor: return codersdk.ValidationError

Instead of returning an untyped error, we instead return a typed error.
This will allow matching on the error at the call site, allowing it to
be dispayed nicely to the user.

DanielleMaywood committed:
chore: appease linter, add todo

DanielleMaywood committed:
fix: add validation error for duplicate slug names

DanielleMaywood commented on coderd/agentapi/subagent_test.go:
```diff
@@ -164,6 +165,328 @@ func TestSubAgentAPI(t *testing.T) {
+			name       string
+			apps       []*proto.CreateSubAgentRequest_App
+			expectApps []database.WorkspaceApp
+			shouldErr  bool
```
Have updated the logic a little bit and return `codersdk.ValidationError` responses now.

johnstcn commented:
> As the new types added are all under the `CreateSubAgentRequest` message, and that message was added in the unreleased version `2.6`, I had assumed it didn't need to be added. I'm happy to make the change though if needed?

@DanielleMaywood fine with me 👍

DanielleMaywood committed:
chore: remove TODO and AsSystemRestricted

DanielleMaywood committed:
Merge branch 'main' into dm-sub-agent-workspace-apps

DanielleMaywood committed:
chore: oops, forgot to add ActionUpdate to permissions

mafredri reviewed (commented):
I think we should utilize transactions here, and I'm a bit torn whether or not to error for app creation. And if we don't error, how do we surface the problem? (Perhaps via some field on the response.)

DanielleMaywood committed:
chore: wrap in a transaction

Wrap the calls to InsertWorkspaceAgent and InsertWorkspaceApp in a
transaction so that they either all succeed or none succeed.

DanielleMaywood committed:
refactor: allow agent to be created when apps fail to be

This change allows an agent to still be created and returned, even when
an app fails to be created. This is made possible by returning each
individual app creation error.

DanielleMaywood committed:
chore: appease linter and formatter

DanielleMaywood commented on coderd/agentapi/subagent.go:
```diff
@@ -71,6 +80,84 @@ func (a *SubAgentAPI) CreateSubAgent(ctx context.Context, req *agentproto.Create
+		})
+		if err != nil {
+			return nil, xerrors.Errorf("insert workspace app: %w", err)
+		}
```
Have made it so the creation of an agent will not fail if an app fails to be created 👍

mafredri commented on coderd/agentapi/subagent.go:
```diff
@@ -71,12 +81,112 @@ func (a *SubAgentAPI) CreateSubAgent(ctx context.Context, req *agentproto.Create
+			}
+			appSlugs[slug] = struct{}{}
+
+			health := database.WorkspaceAppHealthDisabled
+			if app.Healthcheck == nil {
+				app.Healthcheck = &agentproto.CreateSubAgentRequest_App_Healthcheck{}
+			}
+			if app.Healthcheck.Url != "" {
+				health = database.WorkspaceAppHealthInitializing
+			}
+
+			sharingLevel := database.AppSharingLevelOwner
+			switch app.GetShare() {
+			case agentproto.CreateSubAgentRequest_App_AUTHENTICATED:
+				sharingLevel = database.AppSharingLevelAuthenticated
+			case agentproto.CreateSubAgentRequest_App_PUBLIC:
+				sharingLevel = database.AppSharingLevelPublic
+			}
+
+			openIn := database.WorkspaceAppOpenInSlimWindow
+			if app.GetOpenIn() == agentproto.CreateSubAgentRequest_App_TAB {
+				openIn = database.WorkspaceAppOpenInTab
+			}
```
Perhaps we should skip the default and have `switch` for each type with a default that returns an error? Thinking about future changes to the API and potential silent breakages/incorrect behavior that will go unnoticed otherwise.

mafredri commented on coderd/agentapi/subagent.go:
```diff
@@ -71,12 +81,112 @@ func (a *SubAgentAPI) CreateSubAgent(ctx context.Context, req *agentproto.Create
+		}()
+		if err != nil {
+			appErr := &agentproto.CreateSubAgentResponse_AppCreationError{
+				Index: int32(i), //nolint:gosec // This would only overflow if we created 2 billion apps.
```
😄

mafredri reviewed (approved):
Only had one suggestion for a bit of defensive programming, but otherwise this LGTM, nice work!

DanielleMaywood committed:
chore: do some more defensive programming

DanielleMaywood committed:
fix: OOPS
